### PR TITLE
K8SPSMDB-404: Add logic for syncing restore with mongos

### PR DESCRIPTION
**Pull request by blez**
[K8SPSMDB-404](https://jira.percona.com/browse/K8SPSMDB-404) [&#10088;?&#10089;](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

### CODE REVIEW

File: pkg/controller/perconaservermongodb/backup.go [hunk 1]
```diff
@@ -144,3 +145,42 @@ func (h *minHeap) Pop() interface{} {
...
+		if k8sErrors.IsNotFound(err) {
+			return false, nil
+		}
+
+		return false, errors.Wrap(err, "get restore list")
+	}
+
+	for _, rst := range restores.Items {
+		if rst.Status.State != api.RestoreStateReady &&
+			rst.Status.State != api.RestoreStateError {
```
Comment (delgod) [inline pkg/controller/perconaservermongodb/backup.go:161]:
```suggestion
			rst.Status.State != api.RestoreStateError && 
			rst.Spec.ClusterName == cr.Name {
```

File: pkg/controller/perconaservermongodb/backup.go [hunk 2]
```diff
@@ -144,3 +145,42 @@ func (h *minHeap) Pop() interface{} {
...
+	if err := r.client.List(context.TODO(), &bcps, &client.ListOptions{Namespace: cr.Namespace}); err != nil {
+		if k8sErrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, errors.Wrap(err, "get backup list")
+	}
+
+	for _, bcp := range bcps.Items {
+		if bcp.Status.State != api.BackupStateReady &&
+			bcp.Status.State != api.BackupStateError {
```
Comment (delgod) [inline pkg/controller/perconaservermongodb/backup.go:180]:
```suggestion
			bcp.Status.State != api.BackupStateError &&
			bcp.Spec.PSMDBCluster == cr.Name {
```

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/controller/perconaservermongodb/balancer.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+			Namespace: cr.Namespace,
+			LabelSelector: labels.SelectorFromSet(map[string]string{
+				"app.kubernetes.io/name":       "percona-server-mongodb",
+				"app.kubernetes.io/instance":   cr.Name,
+				"app.kubernetes.io/managed-by": "percona-server-mongodb-operator",
+				"app.kubernetes.io/part-of":    "percona-server-mongodb",
+				"app.kubernetes.io/component":  "mongos",
+			}),
+		},
+	)
```
Comment (delgod) [inline pkg/controller/perconaservermongodb/balancer.go:70]:
```suggestion
	mongosPods, err := r.getMongosPods(cr)
```

File: pkg/controller/perconaservermongodb/psmdb_controller.go [hunk 4]
```diff
@@ -430,6 +430,11 @@ func (r *ReconcilePerconaServerMongoDB) Reconcile(request reconcile.Request) (re
 		}
 	}
 
+	err = r.doInCaseOfRestore(cr)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "on restore")
+	}
+
```
Comment (delgod) [inline pkg/controller/perconaservermongodb/psmdb_controller.go:437]: I think we should delete stop balancer and delete mongos from `runRestore` function.
Why we run `doInCaseOfRestore` here?

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| limits | passed |
| demand-backup | passed |
| scaling | passed |
| security-context | passed |
| arbiter | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| rs-shard-migration | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/575/commits/0a15f89d073ed7532576c7b6b267ec9bc9b38522
image: `perconalab/percona-server-mongodb-operator:PR-575-0a15f89d`

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| self-healing | passed |
| limits | passed |
| arbiter | passed |
| scaling | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| security-context | passed |
| scheduled-backup | passed |
| rs-shard-migration | passed |
| liveness | passed |
| smart-update | passed |
| demand-backup-sharded | passed |
| version-service | passed |
| upgrade | passed |
| users | passed |
| upgrade-sharded | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/575/commits/de1568a97ef41bc222d4c6424d75a85c6a705a13
image: `perconalab/percona-server-mongodb-operator:PR-575-de1568a9`

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| smart-update | passed |
| upgrade | passed |
| version-service | passed |
| upgrade-sharded | passed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/575/commits/b5aba832b19f820b547a5722cebdf027e73b13eb
image: `perconalab/percona-server-mongodb-operator:PR-575-b5aba832`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| upgrade | passed |
| smart-update | passed |
| upgrade-sharded | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/575/commits/be788e8e0c05b9a913a4584d3c0cd1b3664d6994
image: `perconalab/percona-server-mongodb-operator:PR-575-be788e8e`

**Review by qJkee — APPROVED** [review]

**Pull request merged**